File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    n = len(maps)
    m = len(maps[0])
    
    dist = [[-1] * m for _ in range(n)]
    dir = [(-1,0),(1,0),(0,-1),(0,1)]
    
    q = deque()
    q.append((0,0))
    dist[0][0] = 1
    while q:
        cur_x, cur_y = q.popleft()
        cur_d = dist[cur_x][cur_y]
        for dx, dy in dir:
            nx, ny = cur_x + dx, cur_y + dy
            if 0<=nx<n and 0<=ny<m:
                if maps[nx][ny] == 1 and dist[nx][ny] < 0:
                    q.append((nx,ny))
                    dist[nx][ny] = cur_d + 1
    return dist[n-1][m-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
#include <tuple>
using namespace std;

int solution(vector<vector<int> > maps)
{
    int answer = -1;
    int n = maps.size();
    int m = maps[0].size();
    
    int dirx[4] = {1,-1,0,0};
    int diry[4] = {0,0,1,-1};
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    queue<tuple<int,int,int>> q;
    q.push({0,0,0});
    visited[0][0] = true;
    while(!q.empty()) {
        auto [curx,cury,curd] = q.front();
        q.pop();
        if (curx == n-1 && cury == m-1) {
            answer = curd + 1;
            break;
        }
        for(int i=0;i<4;i++) {
            int nextx = curx + dirx[i];
            int nexty = cury + diry[i];
            if(0<=nextx && nextx<n && 0<=nexty && nexty<m) {
                if(maps[nextx][nexty] > 0 && !visited[nextx][nexty]) {
                    q.push({nextx,nexty,curd+1});
                    visited[nextx][nexty] = true;
                }
            }
        }
    }
    
    return answer;
}